### PR TITLE
Revert Reader and Writer classes

### PR DESCRIPTION
**Pull request by jbiggsets**
## TL;DR

This PR reverts all of the `Reader` and `Writer` classes that were updated to use `pandas` back to their original structure. The exceptions are the CSV and TSV `Reader` and `Writer` classes that will continue to use `pandas`. 

### Explanation

In a previous PR (#475), we updated all of the `Reader` and `Writer` classes to use `pandas`, rather than reading row-by-row. However, we recently decided to revert most of these changes, and keep _only_ the updates to specific flat file `Reader` and `Writer` classes (CSV and TSV). This was primarily driven by concerns over converting between sparse and dense formats when reading and writing `FeatureSets`, especially with `NDJ`, `DictList`, and `LibSVM` files. (See issue #494 and #493 for some explanation.)

Converting between sparse and dense formats is less of a concern with CSV and TSV `Reader` and `Writer` classes. However, users should still be aware that we are trading memory usage for speed. 

### Benchmarking

Below are some benchmarking results to highlight the tradeoff between speed and memory usage with the new CSV and TSV reader classes. (For more detailed benchmarking results, see [here](https://github.com/EducationalTestingService/skll/files/3637196/test_skll.pdf).)

#### Readers
 - `Reader`: New reader class
 - `Reader2`: Old reader class

####  1000 samples; 20 features
```
%memit fs = Reader.for_path('../test/test1.csv').read()
%memit fs = Reader2.for_path('../test/test1.csv').read()
peak memory: 127.95 MiB, increment: 4.88 MiB
peak memory: 127.29 MiB, increment: -0.42 MiB
```
```
%time fs = Reader.for_path('../test/test1.csv').read()
%time fs = Reader2.for_path('../test/test1.csv').read()
CPU times: user 50.6 ms, sys: 5.31 ms, total: 55.9 ms - Wall time: 53.7 ms
CPU times: user 187 ms, sys: 1.67 ms, total: 189 ms - Wall time: 189 ms
```

#### Writers
 - `Writer`: New writer class
 - `Writer2`: Old writer class

#### 1000 samples; 20 features

```
fs = Reader.for_path('../test/test1.csv').read()

%memit Writer.for_path('../test/_testout.csv', fs).write()
%memit Writer2.for_path('../test/_testout.csv', fs).write()
peak memory: 322.39 MiB, increment: 1.96 MiB
peak memory: 321.96 MiB, increment: 0.02 MiB
```
```
%time Writer.for_path('../test/_testout.csv', fs).write()
%time Writer2.for_path('../test/_testout.csv', fs).write()
CPU times: user 39.5 ms, sys: 5.17 ms, total: 44.6 ms - Wall time: 43.1 ms
CPU times: user 614 ms, sys: 4 ms, total: 618 ms - Wall time: 618 ms
```

### Update (9/27)

One final note: In reverting the `NDJWriter` class back to the original format, I noticed a bug. Essentially, in certain cases, the labels (`y`) were getting written out to `.jsonlines` files as features, rather than labels. Because of this bug, one of the tests that had been written in the interim (`test_send_warnings_to_log`) was failing when I reverted back to the original class, and had to be updated to use a different `.jsonlines` file.

### CODE REVIEW

**Comment by desilinguist** [comment]
This is fantastic @jbiggsets ! Thanks so much for your continued hard work on this. I will review this over the next week.!

File: skll/data/writers.py [hunk 1]
```diff
@@ -8,13 +8,19 @@
...
 import re
 import sys
 import pandas as pd
+import warnings
+
+from csv import DictWriter
 from decimal import Decimal
 
+from scipy.sparse import issparse
+from six.moves import map
```
Comment (desilinguist) [inline skll/data/writers.py]: Don't need this.

File: skll/data/writers.py [hunk 2]
```diff
@@ -535,39 +537,54 @@ def _write_header(self, feature_set, output_file, filter_features):
                       "{" + ','.join(map(str,
```
Comment (desilinguist) [inline skll/data/writers.py]:
```suggestion
                      "{" + ','.join(list(map(str,
```

File: skll/data/writers.py [hunk 3]
```diff
@@ -535,39 +537,54 @@ def _write_header(self, feature_set, output_file, filter_features):
...
                                          sorted(set(self.feat_set.labels)))) +
```
Comment (desilinguist) [inline skll/data/writers.py]:
```suggestion
                                         sorted(set(self.feat_set.labels))))) +
```

**Comment by jbiggsets** [comment]
> You still have `six` in there which we were only using for python 2 compatibility.

File: tests/test_featureset.py [hunk 4]
```diff
@@ -1066,8 +1068,8 @@ def test_featureset_creation_from_dataframe_with_string_ids():
                            "score": [1, 2],
                            "text": ["a b", "b c"]})
     dftest.set_index("id", inplace=True)
-    test_feat_dict_list = [{'a': 1.0, 'b': 1.0, 'c': 0.0},
-                           {'a': 0.0, 'b': 1.0, 'c': 1.0}]
+    test_feat_dict_list = [{'a': 1.0, 'b': 1.0,},
```
Comment (desilinguist) [inline tests/test_featureset.py]:
```suggestion
    test_feat_dict_list = [{'a': 1.0, 'b': 1.0},
```

File: tests/test_featureset.py [hunk 5]
```diff
@@ -1056,6 +1056,8 @@ def test_writing_ndj_featureset_with_string_ids():
     # read in the written file into a featureset and confirm that the
     # two featuresets are equal
     fs_test2 = NDJReader.for_path(output_path).read()
+    print(fs_test)
+    print(fs_test2)
```
Comment (desilinguist) [inline tests/test_featureset.py]: Remove these `print` statements?

File: skll/data/readers.py [hunk 6]
```diff
@@ -878,67 +939,72 @@ def split_with_quotes(self,
...
-            elif row_series.loc[0, 0] == '@relation':
-                self.relation = row_series.loc[0, 1]
-
-        df.columns = columns
-        return self._parse_dataframe(df, self.id_col, self.label_col)
+        field_names = []
+        # Process ARFF header
+        for line in f:
+            # Process encoding
+            if not isinstance(line, text_type):
```
[the remark this answers is not in the document]
Comment (bndgyawali) [reply skll/data/readers.py]: I thought all these were already removed before, but seems there are still leftovers.

[the remark this answers is not in the document]
Comment (desilinguist) [reply skll/data/readers.py]: These came back because @jbiggsets is trying to revert the changes he made earlier and he just happened to use a branch where your 2.7-removal changes had not been merged in.
